plugins/ltac/g_rewrite.mlg

```diff
@@ -99,30 +99,40 @@ ARGUMENT EXTEND rewstrategy
 
     RAW_PRINTED BY { pr_raw_strategy env sigma }
     GLOB_PRINTED BY { pr_glob_strategy env sigma }
+END
 
-  | [ glob(c) ] -> { StratConstr (c, true) }
-  | [ "<-" constr(c) ] -> { StratConstr (c, false) }
-  | [ "subterms" rewstrategy(h) ] -> { StratUnary (Subterms, h) }
-  | [ "subterm" rewstrategy(h) ] -> { StratUnary (Subterm, h) }
-  | [ "innermost" rewstrategy(h) ] -> { StratUnary(Innermost, h) }
-  | [ "outermost" rewstrategy(h) ] -> { StratUnary(Outermost, h) }
-  | [ "bottomup" rewstrategy(h) ] -> { StratUnary(Bottomup, h) }
-  | [ "topdown" rewstrategy(h) ] -> { StratUnary(Topdown, h) }
-  | [ "id" ] -> { StratId }
-  | [ "fail" ] -> { StratFail }
-  | [ "refl" ] -> { StratRefl }
-  | [ "progress" rewstrategy(h) ] -> { StratUnary (Progress, h) }
-  | [ "try" rewstrategy(h) ] -> { StratUnary (Try, h) }
-  | [ "any" rewstrategy(h) ] -> { StratUnary (Any, h) }
-  | [ "repeat" rewstrategy(h) ] -> { StratUnary (Repeat, h) }
-  | [ rewstrategy(h) ";" rewstrategy(h') ] -> { StratBinary (Compose, h, h') }
-  | [ "(" rewstrategy(h) ")" ] -> { h }
-  | [ "choice" ne_rewstrategy_list(h) ] -> { StratNAry (Choice, h) }
-  | [ "old_hints" preident(h) ] -> { StratHints (true, h) }
-  | [ "hints" preident(h) ] -> { StratHints (false, h) }
-  | [ "terms" constr_list(h) ] -> { StratTerms h }
-  | [ "eval" red_expr(r) ] -> { StratEval r }
-  | [ "fold" constr(c) ] -> { StratFold c }
+GRAMMAR EXTEND Gram
+  GLOBAL: rewstrategy;
+  rewstrategy:
+    [ LEFTA
+      [ h = SELF; ";"; h' = rewstrategy_atom -> { StratBinary (Compose, h, h') }
+      | h = rewstrategy_atom -> { h } ] ]
+  ;
+  rewstrategy_atom:
+    [ RIGHTA
+      [ c = glob -> { StratConstr (c, true) }
+      | "<-"; c = constr -> { StratConstr (c, false) }
+      | IDENT "subterms"; h = SELF -> { StratUnary (Subterms, h) }
+      | IDENT "subterm"; h = SELF -> { StratUnary (Subterm, h) }
+      | IDENT "innermost"; h = SELF -> { StratUnary(Innermost, h) }
+      | IDENT "outermost"; h = SELF -> { StratUnary(Outermost, h) }
+      | IDENT "bottomup"; h = SELF -> { StratUnary(Bottomup, h) }
+      | IDENT "topdown"; h = SELF -> { StratUnary(Topdown, h) }
+      | IDENT "id" -> { StratId }
+      | IDENT "fail" -> { StratFail }
+      | IDENT "refl" -> { StratRefl }
+      | IDENT "progress"; h = SELF -> { StratUnary (Progress, h) }
+      | IDENT "try"; h = SELF -> { StratUnary (Try, h) }
+      | IDENT "any"; h = SELF -> { StratUnary (Any, h) }
+      | IDENT "repeat"; h = SELF -> { StratUnary (Repeat, h) }
+      | IDENT "choice"; h = LIST1 rewstrategy -> { StratNAry (Choice, h) }
+      | IDENT "old_hints"; h = preident -> { StratHints (true, h) }
+      | IDENT "hints"; h = preident -> { StratHints (false, h) }
+      | IDENT "terms"; h = LIST0 constr -> { StratTerms h }
+      | IDENT "eval"; r = red_expr -> { StratEval r }
+      | IDENT "fold"; c = constr -> { StratFold c }
+      | "("; h = rewstrategy; ")" -> { h } ] ]
+  ;
```

Comment:
What about
```suggestion
  rewstrategy:
    [ LEFTA
      [ h = SELF; ";"; h' = rewstrategy_atom -> { StratBinary (Compose, h, h') }
      | h = rewstrategy_atom -> { h } ] ]
  ;
  rewstrategy_choice_atom:
    [ NOA
      [ c = glob -> { StratConstr (c, true) }
      | IDENT "id" -> { StratId }
      | IDENT "fail" -> { StratFail }
      | IDENT "refl" -> { StratRefl }
      | "("; h = rewstrategy; ")" -> { h } ]
  ;
  rewstrategy_atom:
    [ RIGHTA
      | "<-"; c = constr -> { StratConstr (c, false) }
      | IDENT "subterms"; h = SELF -> { StratUnary (Subterms, h) }
      | IDENT "subterm"; h = SELF -> { StratUnary (Subterm, h) }
      | IDENT "innermost"; h = SELF -> { StratUnary(Innermost, h) }
      | IDENT "outermost"; h = SELF -> { StratUnary(Outermost, h) }
      | IDENT "bottomup"; h = SELF -> { StratUnary(Bottomup, h) }
      | IDENT "topdown"; h = SELF -> { StratUnary(Topdown, h) }
      | IDENT "progress"; h = SELF -> { StratUnary (Progress, h) }
      | IDENT "try"; h = SELF -> { StratUnary (Try, h) }
      | IDENT "any"; h = SELF -> { StratUnary (Any, h) }
      | IDENT "repeat"; h = SELF -> { StratUnary (Repeat, h) }
      | IDENT "old_hints"; h = preident -> { StratHints (true, h) }
      | IDENT "hints"; h = preident -> { StratHints (false, h) }
      | IDENT "terms"; h = LIST0 constr -> { StratTerms h }
      | IDENT "eval"; r = red_expr -> { StratEval r }
      | IDENT "fold"; c = constr -> { StratFold c }
      | IDENT "choice"; h = LIST1 rewstrategy_choice_atom -> { StratNAry (Choice, h) }
      | h = rewstrategy_choice_atom -> { h } ] ]
  ;
```
I think this is a more sensible precedence for `choice`.  There is no reason to use choice with a single argument, and when passing multiple arguments, I believe it'd be confusing to fail to parenthesize any strategy with spaces in it.  All changes to external developments should be backwards compatible, and my sense is that `rewrite_strat` is anyway not used all that much, so I expect preparing overlays should be pretty easy.  (I'm happy to do it myself if I can find the time.)

Reply:
I'm fine with your changes. Not being an expert of rewrite_strat, I did not want to take the responsability of a change.

In your suggestion, I would just call `rewstrategy_atom` your `rewstrategy_choice_atom`, and, say, `rewstrategy_prefix` your `rewstrategy_atom`. The printer needs also to be adapted.